packages/@expo/prebuild-config/src/plugins/icons/withIosIcons.ts

```diff
@@ -21,31 +21,88 @@ export const withIosIcons: ConfigPlugin = (config) => {
   ]);
 };
 
-export function getIcons(config: Pick<ExpoConfig, 'icon' | 'ios'>): string | null {
-  // No support for empty strings.
-  return config.ios?.icon || config.icon || null;
+export function getIcons(config: Pick<ExpoConfig, 'icon' | 'ios'>): IOSIcons | string | null {
+  const iosSpecificIcons = config.ios?.icon;
+
+  if (iosSpecificIcons) {
+    // For backwards compatibility, the icon can be a string
+    if (typeof iosSpecificIcons === 'string') {
+      return iosSpecificIcons || config.icon || null;
+    }
+
+    // in iOS 18 introduced the ability to specify dark and tinted icons, which users can specify as an object
+    if (!iosSpecificIcons.any && !iosSpecificIcons.dark && !iosSpecificIcons.tinted) {
+      return config.icon || null;
+    }
+
+    return iosSpecificIcons;
+  }
+
+  if (config.icon) {
+    return config.icon;
+  }
+
+  return null;
 }
 
 export async function setIconsAsync(config: ExpoConfig, projectRoot: string) {
   const icon = getIcons(config);
-  if (!icon) {
+
+  if (
+    !icon ||
+    (typeof icon === 'string' && !icon) ||
+    (typeof icon === 'object' && !icon?.any && !icon?.dark && !icon?.tinted)
+  ) {
     WarningAggregator.addWarningIOS('icon', 'No icon is defined in the Expo config.');
   }
+
   // Something like projectRoot/ios/MyApp/
   const iosNamedProjectRoot = getIosNamedProjectPath(projectRoot);
 
   // Ensure the Images.xcassets/AppIcon.appiconset path exists
   await fs.ensureDir(join(iosNamedProjectRoot, IMAGESET_PATH));
 
+  const imagesJson: ContentsJson['images'] = [];
+
+  const baseIconPath = typeof icon === 'object' ? icon?.any || icon?.dark || icon?.tinted : icon;
+
   // Store the image JSON data for assigning via the Contents.json
-  const imagesJson: ContentsJson['images'] = await generateUniversalIconAsync(projectRoot, {
-    icon,
+  const baseIcon = await generateUniversalIconAsync(projectRoot, {
+    icon: baseIconPath,
     cacheKey: 'universal-icon',
     iosNamedProjectRoot,
     platform: 'ios',
   });
 
-  // Finally, write the Config.json
+  imagesJson.push(baseIcon);
+
+  if (typeof icon === 'object') {
+    if (icon?.dark) {
+      const darkIcon = await generateUniversalIconAsync(projectRoot, {
+        icon: icon.dark,
+        cacheKey: 'universal-icon-dark',
+        iosNamedProjectRoot,
+        platform: 'ios',
+        appearance: 'dark',
+      });
+
+      imagesJson.push(darkIcon);
+    }
+
+    if (icon?.tinted) {
+      const tintedIcon = await generateUniversalIconAsync(projectRoot, {
```

Comment:
tinted icons can have transparent backgrounds but this function removes the transparency from the icon.

https://developer.apple.com/design/human-interface-guidelines/app-icons#Platform-considerations

Reply:
@EvanBacon If I'm reading the guidelines correctly, in addition to the tinted icons (where you provide a fully opaque icon), you also provide a transparent icon for the dark appearance.

So it seems like transparency should be removed for the `any` variant, but preserved for `dark` and `tinted` variants. Is this correct?

Reply:
I think it needs to be removed from `tinted` too. Only preserved in `dark`.

Reply:
Looks like you guys didn't implement this correctly and tinted icons have transparency removed.